src/Compilers/CSharp/Portable/Binder/SwitchExpressionArmBinder.cs

```diff
@@ -33,15 +33,16 @@ public SwitchExpressionArmBinder(SwitchExpressionArmSyntax arm, ExpressionVariab
         internal override BoundSwitchExpressionArm BindSwitchExpressionArm(SwitchExpressionArmSyntax node, DiagnosticBag diagnostics)
         {
             Debug.Assert(node == _arm);
-            var caseBinder = this.GetBinder(node);
-            var hasErrors = _switchExpressionBinder.SwitchGoverningType.IsErrorType();
-            var locals = _armScopeBinder.Locals;
-            var pattern = caseBinder.BindPattern(node.Pattern, _switchExpressionBinder.SwitchGoverningType, hasErrors, diagnostics);
-            var guard = node.WhenClause != null
-                ? caseBinder.BindBooleanExpression(node.WhenClause.Condition, diagnostics)
+            Binder armBinder = this.GetBinder(node);
+            bool hasErrors = _switchExpressionBinder.SwitchGoverningType.IsErrorType();
+            ImmutableArray<LocalSymbol> locals = _armScopeBinder.Locals;
+            BoundPattern pattern = armBinder.BindPattern(node.Pattern, _switchExpressionBinder.SwitchGoverningType, hasErrors, diagnostics);
+            BoundExpression guard = node.WhenClause != null
+                ? armBinder.BindBooleanExpression(node.WhenClause.Condition, diagnostics)
                 : null;
-            var result = caseBinder.BindValue(node.Expression, diagnostics, BindValueKind.RValue);
-            return new BoundSwitchExpressionArm(node, locals, pattern, guard, result, hasErrors);
+            BoundExpression armResult = armBinder.BindValue(node.Expression, diagnostics, BindValueKind.RValue);
+            var label = new GeneratedLabelSymbol("arm");
+            return new BoundSwitchExpressionArm(node, locals, pattern, guard, armResult, label, hasErrors | pattern.HasErrors);
```

Comment:
>| pattern.HasErrors) [](start = 106, length = 20)

Is it necessary to include `pattern.HasErrors`? And if so, why not `guard.HasErrors` or `armResult.HasErrors`?

Reply:
When the pattern has errors or is subsumed, we still want to treat the arm expression as OK for flow-analysis purposes. Errors in the pattern can cause it to be dropped from the decision dag (e.g. when subsumed). By placing the error flag on the arm flow analysis can use it to act in such a way as to minimize cascaded diagnostics. For the switch statement we treat it as reachable so that we do not report unreachable warnings. We don't handle the switch expression specially now, but the intent is to treat this as *unreachable* for flow analysis so that we do not report any spurious issues with unassigned variables. This flag is setting the stage for that.

---
In reply to: [172630166](https://github.com/dotnet/roslyn/pull/24059#discussion_r172630166) [](ancestors = 172630166)